tools/restatectl/src/connection.rs

```diff
@@ -0,0 +1,267 @@
+
+        let iterator = match role {
+            Some(role) => Either::Left(nodes_config.iter_role(role)),
+            None => Either::Right(nodes_config.iter()),
+        }
+        .sorted_by(|left, right| {
+            // nodes that we already have channels open for, gets higher presence.
+            match (
+                channels.contains_key(&left.0),
+                channels.contains_key(&right.0),
+            ) {
+                (true, false) => Ordering::Less,
+                (false, true) => Ordering::Greater,
+                (_, _) => left.0.cmp(&right.0),
+            }
+        });
+
+        let mut errors = NodesErrors::default();
+
+        for (node_id, node) in iterator {
+            // avoid creating new channels on each iteration. Instead cheaply copy the channels
+            let channel = channels
+                .entry(node_id)
+                .or_insert_with(|| grpc_channel(node.address.clone()));
+
+            let result = closure(channel.clone()).await;
+            match result {
+                Ok(response) => return Ok(response),
+                Err(status) => {
+                    errors.error(node.address.clone(), status);
+                }
+            }
+        }
+
+        if errors.is_empty() {
+            Err(ConnectionInfoError::NoAvailableNodes(NoRoleError(role)))
+        } else {
+            Err(ConnectionInfoError::NodesErrors(errors))
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ConnectionInfoError {
+    #[error("Could not retrieve nodes configuration. Possible un provisioned cluster!")]
+    MissingNodesConfiguration,
+
+    #[error("Failed to decode nodes configuration from node {0}: {1}")]
+    InvalidNodesConfiguration(AdvertisedAddress, StorageDecodeError),
+
+    #[error(transparent)]
+    NodesErrors(NodesErrors),
+
+    #[error(transparent)]
+    NoAvailableNodes(NoRoleError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub struct NoRoleError(Option<Role>);
+
+impl Display for NoRoleError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.0 {
+            Some(role) => {
+                write!(f, "No available {role} nodes to satisfy the request")?;
+            }
+            None => {
+                write!(f, "No available nodes to satisfy the request")?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct NodesErrors {
+    node_status: Vec<(AdvertisedAddress, Status)>,
+}
+
+impl NodesErrors {
+    fn error(&mut self, node: AdvertisedAddress, status: Status) {
+        self.node_status.push((node, status));
+    }
+
+    fn is_empty(&self) -> bool {
+        self.node_status.is_empty()
+    }
+}
+
+impl Display for NodesErrors {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(
+            f,
+            "Encountered multiple errors trying to retrieve nodes configurations:"
```

Comment:
```suggestion
            "Encountered multiple errors trying to retrieve nodes' configuration:"
```

Reply:
We seem to use `NodesErrors` also for the `try_each` call where we are not necessarily trying to retrieve the `NodesConfiguration`. Maybe this needs an update of the `Display` implementation.

Reply:
Yes, makes sense